tree/tree/src/TTree.cxx

```diff
@@ -4802,6 +4814,53 @@ Int_t TTree::FlushBaskets() const
    Int_t nerror = 0;
    TObjArray *lb = const_cast<TTree*>(this)->GetListOfBranches();
    Int_t nb = lb->GetEntriesFast();
+
+#ifdef R__USE_IMT
+   if (ROOT::IsImplicitMTEnabled() && fIMTEnabled) {
+      if (fSortedBranches.empty()) { const_cast<TTree*>(this)->InitializeSortedBranches(); }
+
+      BoolRAIIToggle sentry(fIMTFlush);
+      fIMTZipBytes.store(0);
+      fIMTTotBytes.store(0);
+      std::atomic<Int_t> nerrpar(0);
+      std::atomic<Int_t> nbpar(0);
+      std::atomic<Int_t> pos(0);
+      tbb::task_group g;
```

Comment:
Here's an example of a deadlock that could have happened in older versions of CMSSW:

- CMSSW takes a mutex on the output module.
- CMSSW performs a ROOT call which causes a `FlushBaskets` operation.
- ROOT creates many basket flushing tasks and then `wait` on the task group.
- TBB's runtime notice the thread blocks in `wait` and schedules another CMSSW task on the same thread.
- The other CMSSW also needs to access the output module, hence takes a mutex.
- Since the thread already holds the output module's mutex, trying to acquire it again causes a deadlock.

Now, Chris tells me that CMSSW has switched away from the mutex for output module access - but this might be a good example.

I pushed a long comment for `FlushBaskets` outlining the above.